### ISSUE TITLE
[tcp] All Closer Closures are Buggy / Inefficient / Unneeded

### ISSUE THREAD

**Issue by BrianZill** [demikernel/catnip#109]
Description
========

At present, there are 6 closure routines in closer.rs.  Four of them emit TCP segments, _all_ of which are malformed (two of them don't set the sequence number field, the other two don't set the acknowledgement number field.  They also resume looping after sending those segments, even though the event each of them waits on (a particular TCP state transition) logically only occurs once.  The remaining two routines simply wait for our TCP state to reach a particular value (and merely return an error value when that happens) -- these appear to just be a way to terminate the other closure loops when either of the two end states is reached.

There doesn't appear to be a reason to have any of these routines.  All of them are waiting around for particular state transitions to occur in order to do something, and in all of these cases it appears that the code making the particular state transition could just do that something directly and be more efficient in the process.

**Comment by gatoWololo** [comment]
The `test_connect_disconnect` test under `src/protocols/tcp/tests/established.rs` sets up a connection between a client and server and then immediately disconnects the connection.

This test passes. However, if some more code is added to send some bytes between client and server. The test will fail with a panic: `bad peer state Established`.
```rust
        if header.fin && header.ack {
            match self.state.get() {
                State::FinWait1 => self.state.set(State::TimeWait1),
                s => panic!("bad peer state {:?}", s),
            }
```
I think this is what is happening:
1) Client sends data to server.
2) Server attempts to end the connection by sending a `FIN`. Sever also includes an `ACK` to acknowledge the bytes it received in step 1.
3) Client panics in the code above.

**Comment by BrianZill** [comment]
> The `test_connect_disconnect` test under `src/protocols/tcp/tests/established.rs` sets up a connection between a client and server and then immediately disconnects the connection.
> 
> This test passes. However, if some more code is added to send some bytes between client and server. The test will fail with a panic: `bad peer state Established`.
> 
> ```rust
>         if header.fin && header.ack {
>             match self.state.get() {
>                 State::FinWait1 => self.state.set(State::TimeWait1),
>                 s => panic!("bad peer state {:?}", s),
>             }
> ```
> 
> I think this is what is happening:
> 
> 1. Client sends data to server.
> 2. Server attempts to end the connection by sending a `FIN`. Sever also includes an `ACK` to acknowledge the bytes it received in step 1.
> 3. Client panics in the code above.

Yes, the entire current mechanism for determining when/how to send a FIN is badly busted.  I'm fixing it to do it the right way and will be removing all of the closer closures soon.

**Comment by BrianZill** [comment]
Fixed in PR #135.
